{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Methylysis\n",
    "Methylysis is a tool to analyze the DNA methylation data for two different Illumina arrays, 450K and 850K array. The current tool uses two most used well-defined pipelines, named The Chip Analysis Methylation Pipeline [ChAMP](https://bioconductor.org/packages/release/bioc/vignettes/ChAMP/inst/doc/ChAMP.html#section-svd-plot) and [minfi](https://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.html) pipeline.\n",
    "We introduced different input options for the user to run both pipelines in a more customized way. Based on the above-mentioned pipelines, we selected the user's options. Users can run either pipeline of their choice with different filteration criteria, and of course they can use two different runs and compare the results by themselves using downstream processing, like Venn analysis.\n",
    "\n",
    "*Please note*\n",
    "1. After loading the sample directory, the pipeline will start immediately by displaying the notification <font color='blue'>\"Computing methylysis, please wait...\"</font>. When the notification turns off, the user can go to different tabs to display the result. Please wait 10-15 seconds to display the result on the tab. Depending on the sample size, it may requires more time to display properly.\n",
    "2. If the methylysis page goes **dim**, the analysis may encounter some errors during the run and the program stops working. Please refresh the page and run the same analysis again with different filteration criteria. Example, for ChAMP pipeline, if the user select the  \"adjusted P-value\" = 0.05 (as default), may be for the sample data, there is no differentially methylated CpGs at that value. Please change the adjusted P-value (recommend to set it at 1, and check the table after the run that what is appropriate cut-off) and run the pipeline again.\n",
    "\n",
    "## How to use\n",
    "Details are provided below -\n",
    "\n",
    "### Data upload & Parameters setup\n",
    "The current version can handle the upload of the data directory. Please put all RAW idat files (as generated by the Illumina sequencer) and the \"Sample_sheet.csv\" together in a directory.\n",
    "#### Structure of sample_sheet.csv\n",
    "The Sample_sheet.csv must have the following components -\n",
    "1. Sample_Name\n",
    "2. Sample_Group\n",
    "3. Sentrix_ID\n",
    "4. Sentrix_position\n",
    "\n",
    "*Note* If the user uses Microsoft Excel to build the Sample_sheet.csv, please check that\n",
    "1. Sentrix_ID : are in text format (not in number format, which Excel will change to scientific numbers and will not properly displayed).\n",
    "2. Optional check: Copy and paste the Excel table of Sample_sheet in some text editor like notepad or VS code and check the format.\n",
    "\n",
    "To download a template of Sample_sheet.csv, please [check here]().\n",
    "\n",
    "#### Parameters setup\n",
    "\n",
    "##### Choose analysis algorithm\n",
    "Currently, we have included two most usable algorithms to analyze the data - ChAMP and minfi.\n",
    "![algorithm choice](images/methylysis1.png)\n",
    "\n",
    "###### ChAMP pipeline parameters\n",
    "\n",
    "1. *Choose type of Illumina array*: Two options are provided to choose, namely EPIC/850K array and 450K array from Illumina array analysis.\n",
    "\n",
    "2. *Adjusted P-value*: User can define there own adjusted P-value to run the analysis. The default is 0.05.\n",
    "\n",
    "3. *Normalization*: User can choose different normalization methods from the drop-down list,\n",
    "    - BMIQ (Beta-Mixture Quantile Normalization) {cite:p}`teschendorff2013beta`,\n",
    "    - SWAN (Sub-Quantile Within Arra) {cite:p}`touleimat2012complete`,{cite:p}`maksimovic2012swan`,\n",
    "    - PBC (Peak-Based Correction) {cite:p}`dedeurwaerder2011evaluation`,\n",
    "    - FunctionalNormalization {cite:p}`fortin2014functional`.\n",
    "The default setup will run with the BMIQ normalization method. Please check references for different type of normalization method\n",
    "\n",
    "4. *Batch Effect Correction*: *ComBat* function is used to correct the batch effect. User can choose whether to compute the batch effect or not by clicking the button. When the button is <font color='green'>green</font> (ON), then the pipeline will compute the batch effect and if the button is <font color='red'>red</font> (OFF), the pipeline will continue without analyzing the batch effect. Please check the reference for the batch effect correction using combat method {cite:p}`johnson2007adjusting`.\n",
    "\n",
    "5. *Cell Type Heterogeneity*: Houseman et al (2013){cite:p}`houseman2012dna` algorithm is applied to calculate the cell-type heterogeneity from PBMC dataset using the *refbase* function. This is deactivated as default. Press the button to activate and run during the analysis.\n",
    "![champ](images/methylysis2.png)\n",
    "\n",
    "###### minfi pipeline parameters\n",
    "\n",
    "1. *Choose preprocess method* : There are several methods available for preprocessing or normalizing the raw data using the RGset. Here we listed them as the user's input options to select the preprocess/normalization method as per their choice:\n",
    "    - Raw: No processing of the raw data,\n",
    "    - SWAN: Subset Quantile Within array Normalization {cite:p}`maksimovic2012swan`,{cite:p}`touleimat2012complete`,\n",
    "    - Qunatile: Quantile normalization method {cite:p}`maksimovic2012swan`,\n",
    "    - Noob: Noob preprocessing {cite:p}`triche2013low`,\n",
    "    - Illumina: llumina preprocessing, as performed by Genome Studio (reverse engineered by minfi authors) {cite:p}`aryee2014minfi`\n",
    "    - Funnorm: Funcational normalization {cite:p}`fortin2014functional`\n",
    "\n",
    "2. *Select filteration method*: In this section, we assigned options for the user to perform the different filters, like removal of XY chormosomes from the analysis, removal of SNPs or removal of non-specific probes from the dataset. By default, the pipeline will use p-value detection 0.01.\n",
    "\n",
    "3. *Compute cell type heterogeneity*: similar as ChAMP, we used Houseman method to correct the cell type heterogeneity. In minfi pipeline we used default minfi function *estimateCellCounts*. As before, user can choose to avoid this if the samples are not from PBMC cell types.\n",
    "\n",
    "4. *Please select the compare groups*: In this option, the pipeline will ask the user to type the group names which will be compared for the differential methylation analysis. To compute the ContrastsMatrix, these group names will be used. Please use the same groups as provide in the Sample_sheet.csv.\n",
    "\n",
    "5. *Choose genome annotation database*: To annotate the DMC list from the analysis, use the human genome reference annotation data file. For 850K array, make sure to use the hg38 array to compare the result with the output of ChAMP pipeline.\n",
    "![minfi](images/methylysis3.png)\n",
    "\n",
    "###### Technical setup:\n",
    "1. *Number of cores*: Both pipeline can run on 1 core which will take more time to compute the entire process. User can choose to setup the number of cores depending the availability. Please select maximum (n-1) cores (where n is the number of total cores in the computer).\n",
    "\n",
    "2. *Data upload*\n",
    "The user should set the parameters first and choose all parameters as described above and then upload the data directory. As soon as the pipeline finishes the upload of the data directory, it will start running the analysis.\n",
    "![data upload](images/methylysis4.png)\n",
    "\n",
    "##### Requirement for data upload\n",
    "1. idat files: all idat files, green and red as received from Illumina sequencing array should be provided for the analysis. All files should be in one directory/folder.\n",
    "2. Sample_sheet.csv: the \"Sample_sheet.csv\" should also be provided in the same directory with idat files and"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Analysis Result\n",
    "We implemented two different appraoches to save/view the analysis result.\n",
    "A. First, all results from the analysis pipeline will be saved in the same folder from where the data loaded, and\n",
    "B. Second, using the on-screen visualization tool.\n",
    "\n",
    "In the visaulization tabs of methylysis tool, we introduced 4 tabs for differential methylation analysis result,\n",
    "1. **QC result**: QC result from each pipeline mentioned above will be generated and presented here in PDF format. User can download the PDF, however the result will also be saved in the data directory.\n",
    "2. **Normalized data table**: Only first 100 rows from the normalized \\beta values will appear in the visualization tab. The full result will be saved on the data directory.\n",
    "3. **Cell type deconvolution plot**: If user choose to analyze the cell type deconvolution, the analysis figure will appear here. Please note, current version only supports the PBMC deconvolution.\n",
    "4. **DMC table**: First 100 rows of the differentially methylated CpGs will appear here and the full result will be saved as mentioned above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## R packages used\n",
    "1. ChAMP bioconductor\n",
    "2. minfi bioconductor\n",
    "3. IlluminaHumanMethylation450kanno.ilmn12.hg19\n",
    "4. IlluminaHumanMethylationEPICanno.ilm10b5.hg38\n",
    "5. IlluminaHumanMethylation450kmanifest\n",
    "6. RColorBrewer\n",
    "7. missMethyl\n",
    "8. matrixStats\n",
    "9. minfiData\n",
    "10. Gviz\n",
    "11. DMRcate\n",
    "12. stringr\n",
    "13. RCurl\n",
    "14. Biobase"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}