{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to convert .Rmd to .ipynb\n",
    "1. Require an R package, rmd2jupyter\n",
    "```\n",
    "devtools::install_github(\"mkearney/rmd2jupyter\")\n",
    "library(rmd2jupyter)\n",
    "```\n",
    "2. Prepare the .Rmd file\n",
    "3. Convert the file\n",
    "```\n",
    "rmd2jupyter(\"sample_file.Rmd\")\n",
    "```\n",
    "\n",
    "\n",
    "## Notes\n",
    "\n",
    "\n",
    "[1] See Table 1\n",
    "\n",
    "[2] adding other parameters to the plot to make it more useful and\n",
    "understandable\n",
    "\n",
    "[3] In this, the value is multiplied by 100\n",
    "\n",
    "[4] For strings parallel to the axis, adj =0 means left or bottom\n",
    "alignment and adj = 1 means right or top alignment\n",
    "\n",
    "[5] [R plot **pch**\n",
    "symbols](http://www.sthda.com/english/wiki/r-plot-pch-symbols-the-different-point-shapes-available-in-r)\n",
    "\n",
    "[6] could not find the required symbols in LaTeX\n",
    "\n",
    "[7] convert data into a local data frame,\n",
    "\n",
    "[8] if library(dplyr) has been called and filter with grepl\n",
    "\n",
    "[9] **Note:** *par* function, par sets or adjusts plotting parameters,\n",
    "and *mar* is a numeric vector of length 4, mar(bottom, left, top, right)\n",
    "\n",
    "[10] mgp is a numeric vector of length 3, sets the axis label locations\n",
    "relative to the edge of the inner plot window. The first value\n",
    "represents the location of the labels, the second - the tick-maeks\n",
    "labels, and third - the tick marks. The default is c(3,1,0).\n",
    "\n",
    "[11] is also a numeric value indicating the orientation of the tickmarks\n",
    "labels and any other text added to the plot after the initialization. 0\n",
    "- always parallel to the axis (default), 1 - always horizontal, 2 -\n",
    "always perpendiclar to the axis, and 3 - always vertical\n",
    "\n",
    "[12] ifelse is used as if ... else statement. ifelse(text expression, x,\n",
    "y)\n",
    "\n",
    "[13] Abbreviations used$:$ Con$:$Control and Exp$:$ Exposed\n",
    "\n",
    "[14] TB patient, household contacts and controls data from Link√∂ping,\n",
    "Sweden\n",
    "\n",
    "[15] this can only be used if the sample group contains more than two\n",
    "groups. In this case, we used the patient sample data as \"Exposed\"\n",
    "because one data sample group cannot be analyzed. We have only two\n",
    "sample groups, Exposed vs NonExposed\n",
    "\n",
    "[16] [HumanMethylation450 manifest\n",
    "file](https://emea.support.illumina.com/downloads/infinium_humanmethylation450_product_files.html)\n",
    "\n",
    "[17] [Links to source\n",
    "package](https://github.com/kevinblighe/EnhancedVolcano)\n",
    "\n",
    "[18] [Links to github package](https://wencke.github.io/)\n",
    "\n",
    "[19] A. A list of 4131 DMGs (BH adjusted *p value* $<$ 0.05) are used\n",
    "from the CD3 cell types to analyze the functional enrichment.  \n",
    "B. For HLA-DR cell types, a total of 9266 DMGs (BH adjusted *p value*\n",
    "$<$ 0.05) are taken.  \n",
    "C. The functional enrichment analysis of PBMC cell types contain 537\n",
    "DMGs (BH adjusted *p value* $<$ 0.2)\n",
    "\n",
    "[20] [Links to Panther database](http://www.pantherdb.org/)\n",
    "\n",
    "[21] [Links to reactome](https://reactome.org/)\n",
    "\n",
    "[22] [Links to database](http://www.webgestalt.org/)\n",
    "\n",
    "[23] Kyoto Encyclopedia of Gene and Genome\n",
    "\n",
    "[24] Protein ANalysis THrough Evolutionary Relationships\n",
    "\n",
    "[25] WEB-based GEne SeT AnaLysis Toolkit[Link to\n",
    "database](http://www.webgestalt.org/)\n",
    "\n",
    "[26] For the dotplot, I used *ggpubr* package in R"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}