
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>1. Methylysis &#8212; DNA Methylation Data Analysis</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2. Multi-D Analysis" href="multid.html" />
    <link rel="prev" title="Welcome to methylR: DNA Methylation Data Analysis Pipeline" href="intro.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo-final.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">DNA Methylation Data Analysis</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Welcome to methylR: DNA Methylation Data Analysis Pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  DNA methylation
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   1. Methylysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Feature analyses
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="multid.html">
   2. Multi-D Analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="genomicfeatures.html">
   3. Gene Features analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pairwise.html">
   4. Pairwise &amp; Heatmap analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="volcano.html">
   5. Volcano plot
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Association study
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="go.html">
   6. Gene Ontology (GO) enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pathwayanalysis.html">
   7. Pathway enrichment analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Set analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="vennanalysis.html">
   8. Venn analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="upset.html">
   9. UpSet Plots
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  References
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="markdown.html">
   10. Citations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Appendices
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks.html">
   11. How to convert .Rmd to .ipynb
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/methylysis.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/JD2112/methylr"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/JD2112/methylr/issues/new?title=Issue%20on%20page%20%2Fmethylysis.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#how-to-use">
   1.1. How to use
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-upload-parameters-setup">
     1.1.1. Data upload &amp; Parameters setup
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#structure-of-sample-sheet-csv">
       1.1.1.1. Structure of sample_sheet.csv
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#parameters-setup">
       1.1.1.2. Parameters setup
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#choose-analysis-algorithm">
         1.1.1.2.1. Choose analysis algorithm
        </a>
        <ul class="nav section-nav flex-column">
         <li class="toc-h6 nav-item toc-entry">
          <a class="reference internal nav-link" href="#champ-pipeline-parameters">
           1.1.1.2.1.1. ChAMP pipeline parameters
          </a>
         </li>
         <li class="toc-h6 nav-item toc-entry">
          <a class="reference internal nav-link" href="#minfi-pipeline-parameters">
           1.1.1.2.1.2. minfi pipeline parameters
          </a>
         </li>
         <li class="toc-h6 nav-item toc-entry">
          <a class="reference internal nav-link" href="#technical-setup">
           1.1.1.2.1.3. Technical setup:
          </a>
         </li>
        </ul>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#requirement-for-data-upload">
         1.1.1.2.2. Requirement for data upload
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analysis-result">
   1.2. Analysis Result
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#r-packages-used">
   1.3. R packages used
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Methylysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#how-to-use">
   1.1. How to use
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-upload-parameters-setup">
     1.1.1. Data upload &amp; Parameters setup
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#structure-of-sample-sheet-csv">
       1.1.1.1. Structure of sample_sheet.csv
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#parameters-setup">
       1.1.1.2. Parameters setup
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#choose-analysis-algorithm">
         1.1.1.2.1. Choose analysis algorithm
        </a>
        <ul class="nav section-nav flex-column">
         <li class="toc-h6 nav-item toc-entry">
          <a class="reference internal nav-link" href="#champ-pipeline-parameters">
           1.1.1.2.1.1. ChAMP pipeline parameters
          </a>
         </li>
         <li class="toc-h6 nav-item toc-entry">
          <a class="reference internal nav-link" href="#minfi-pipeline-parameters">
           1.1.1.2.1.2. minfi pipeline parameters
          </a>
         </li>
         <li class="toc-h6 nav-item toc-entry">
          <a class="reference internal nav-link" href="#technical-setup">
           1.1.1.2.1.3. Technical setup:
          </a>
         </li>
        </ul>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#requirement-for-data-upload">
         1.1.1.2.2. Requirement for data upload
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analysis-result">
   1.2. Analysis Result
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#r-packages-used">
   1.3. R packages used
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="methylysis">
<h1><span class="section-number">1. </span>Methylysis<a class="headerlink" href="#methylysis" title="Permalink to this headline">¶</a></h1>
<p>Methylysis is a tool to analyze the DNA methylation data for two different Illumina arrays, 450K and 850K array. The current tool uses two most used well-defined pipelines, named The Chip Analysis Methylation Pipeline <a class="reference external" href="https://bioconductor.org/packages/release/bioc/vignettes/ChAMP/inst/doc/ChAMP.html#section-svd-plot">ChAMP</a> and <a class="reference external" href="https://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.html">minfi</a> pipeline.
We introduced different input options for the user to run both pipelines in a more customized way. Based on the above-mentioned pipelines, we selected the user’s options. Users can run either pipeline of their choice with different filteration criteria, and of course they can use two different runs and compare the results by themselves using downstream processing, like Venn analysis.</p>
<p><em>Please note</em></p>
<ol class="simple">
<li><p>After loading the sample directory, the pipeline will start immediately by displaying the notification <font color='blue'>”Computing methylysis, please wait…”</font>. When the notification turns off, the user can go to different tabs to display the result. Please wait 10-15 seconds to display the result on the tab. Depending on the sample size, it may requires more time to display properly.</p></li>
<li><p>If the methylysis page goes <strong>dim</strong>, the analysis may encounter some errors during the run and the program stops working. Please refresh the page and run the same analysis again with different filteration criteria. Example, for ChAMP pipeline, if the user select the  “adjusted P-value” = 0.05 (as default), may be for the sample data, there is no differentially methylated CpGs at that value. Please change the adjusted P-value (recommend to set it at 1, and check the table after the run that what is appropriate cut-off) and run the pipeline again.</p></li>
</ol>
<div class="section" id="how-to-use">
<h2><span class="section-number">1.1. </span>How to use<a class="headerlink" href="#how-to-use" title="Permalink to this headline">¶</a></h2>
<p>Details are provided below -</p>
<div class="section" id="data-upload-parameters-setup">
<h3><span class="section-number">1.1.1. </span>Data upload &amp; Parameters setup<a class="headerlink" href="#data-upload-parameters-setup" title="Permalink to this headline">¶</a></h3>
<p>The current version can handle the upload of the data directory. Please put all RAW idat files (as generated by the Illumina sequencer) and the “Sample_sheet.csv” together in a directory.</p>
<div class="section" id="structure-of-sample-sheet-csv">
<h4><span class="section-number">1.1.1.1. </span>Structure of sample_sheet.csv<a class="headerlink" href="#structure-of-sample-sheet-csv" title="Permalink to this headline">¶</a></h4>
<p>The Sample_sheet.csv must have the following components -</p>
<ol class="simple">
<li><p>Sample_Name</p></li>
<li><p>Sample_Group</p></li>
<li><p>Sentrix_ID</p></li>
<li><p>Sentrix_position</p></li>
</ol>
<p><em>Note</em> If the user uses Microsoft Excel to build the Sample_sheet.csv, please check that</p>
<ol class="simple">
<li><p>Sentrix_ID : are in text format (not in number format, which Excel will change to scientific numbers and will not properly displayed).</p></li>
<li><p>Optional check: Copy and paste the Excel table of Sample_sheet in some text editor like notepad or VS code and check the format.</p></li>
</ol>
<p>To download a template of Sample_sheet.csv, please <span class="xref myst">check here</span>.</p>
</div>
<div class="section" id="parameters-setup">
<h4><span class="section-number">1.1.1.2. </span>Parameters setup<a class="headerlink" href="#parameters-setup" title="Permalink to this headline">¶</a></h4>
<div class="section" id="choose-analysis-algorithm">
<h5><span class="section-number">1.1.1.2.1. </span>Choose analysis algorithm<a class="headerlink" href="#choose-analysis-algorithm" title="Permalink to this headline">¶</a></h5>
<p>Currently, we have included two most usable algorithms to analyze the data - ChAMP and minfi.
<img alt="algorithm choice" src="_images/methylysis1.png" /></p>
<div class="section" id="champ-pipeline-parameters">
<h6><span class="section-number">1.1.1.2.1.1. </span>ChAMP pipeline parameters<a class="headerlink" href="#champ-pipeline-parameters" title="Permalink to this headline">¶</a></h6>
<ol class="simple">
<li><p><em>Choose type of Illumina array</em>: Two options are provided to choose, namely EPIC/850K array and 450K array from Illumina array analysis.</p></li>
<li><p><em>Adjusted P-value</em>: User can define there own adjusted P-value to run the analysis. The default is 0.05.</p></li>
<li><p><em>Normalization</em>: User can choose different normalization methods from the drop-down list,</p>
<ul class="simple">
<li><p>BMIQ (Beta-Mixture Quantile Normalization) <span id="id1">[<a class="reference internal" href="markdown.html#id8" title="Andrew E Teschendorff, Francesco Marabita, Matthias Lechner, Thomas Bartlett, Jesper Tegner, David Gomez-Cabrero, and Stephan Beck. A beta-mixture quantile normalization method for correcting probe design bias in illumina infinium 450 k dna methylation data. Bioinformatics, 29(2):189–196, 2013.">TML+13</a>]</span>,</p></li>
<li><p>SWAN (Sub-Quantile Within Arra) <span id="id2">[<a class="reference internal" href="markdown.html#id10" title="Nizar Touleimat and Jörg Tost. Complete pipeline for infinium® human methylation 450k beadchip data processing using subset quantile normalization for accurate dna methylation estimation. Epigenomics, 4(3):325–341, 2012.">TT12</a>]</span>,<span id="id3">[<a class="reference internal" href="markdown.html#id51" title="Jovana Maksimovic, Lavinia Gordon, and Alicia Oshlack. Swan: subset-quantile within array normalization for illumina infinium humanmethylation450 beadchips. Genome biology, 13(6):1–12, 2012.">MGO12</a>]</span>,</p></li>
<li><p>PBC (Peak-Based Correction) <span id="id4">[<a class="reference internal" href="markdown.html#id9" title="Sarah Dedeurwaerder, Matthieu Defrance, Emilie Calonne, Hélene Denis, Christos Sotiriou, and François Fuks. Evaluation of the infinium methylation 450k technology. Epigenomics, 3(6):771–784, 2011.">DDC+11</a>]</span>,</p></li>
<li><p>FunctionalNormalization <span id="id5">[<a class="reference internal" href="markdown.html#id7" title="Jean-Philippe Fortin, Aurélie Labbe, Mathieu Lemire, Brent W Zanke, Thomas J Hudson, Elana J Fertig, Celia MT Greenwood, and Kasper D Hansen. Functional normalization of 450k methylation array data improves replication in large cancer studies. Genome biology, 15(11):1–17, 2014.">FLL+14</a>]</span>.
The default setup will run with the BMIQ normalization method. Please check references for different type of normalization method</p></li>
</ul>
</li>
<li><p><em>Batch Effect Correction</em>: <em>ComBat</em> function is used to correct the batch effect. User can choose whether to compute the batch effect or not by clicking the button. When the button is <font color='green'>green</font> (ON), then the pipeline will compute the batch effect and if the button is <font color='red'>red</font> (OFF), the pipeline will continue without analyzing the batch effect. Please check the reference for the batch effect correction using combat method <span id="id6">[<a class="reference internal" href="markdown.html#id49" title="W Evan Johnson, Cheng Li, and Ariel Rabinovic. Adjusting batch effects in microarray expression data using empirical bayes methods. Biostatistics, 8(1):118–127, 2007.">JLR07</a>]</span>.</p></li>
<li><p><em>Cell Type Heterogeneity</em>: Houseman et al (2013)<span id="id7">[<a class="reference internal" href="markdown.html#id50" title="Eugene Andres Houseman, William P Accomando, Devin C Koestler, Brock C Christensen, Carmen J Marsit, Heather H Nelson, John K Wiencke, and Karl T Kelsey. Dna methylation arrays as surrogate measures of cell mixture distribution. BMC bioinformatics, 13(1):1–16, 2012.">HAK+12</a>]</span> algorithm is applied to calculate the cell-type heterogeneity from PBMC dataset using the <em>refbase</em> function. This is deactivated as default. Press the button to activate and run during the analysis.
<img alt="champ" src="_images/methylysis2.png" /></p></li>
</ol>
</div>
<div class="section" id="minfi-pipeline-parameters">
<h6><span class="section-number">1.1.1.2.1.2. </span>minfi pipeline parameters<a class="headerlink" href="#minfi-pipeline-parameters" title="Permalink to this headline">¶</a></h6>
<ol class="simple">
<li><p><em>Choose preprocess method</em> : There are several methods available for preprocessing or normalizing the raw data using the RGset. Here we listed them as the user’s input options to select the preprocess/normalization method as per their choice:</p>
<ul class="simple">
<li><p>Raw: No processing of the raw data,</p></li>
<li><p>SWAN: Subset Quantile Within array Normalization <span id="id8">[<a class="reference internal" href="markdown.html#id51" title="Jovana Maksimovic, Lavinia Gordon, and Alicia Oshlack. Swan: subset-quantile within array normalization for illumina infinium humanmethylation450 beadchips. Genome biology, 13(6):1–12, 2012.">MGO12</a>]</span>,<span id="id9">[<a class="reference internal" href="markdown.html#id10" title="Nizar Touleimat and Jörg Tost. Complete pipeline for infinium® human methylation 450k beadchip data processing using subset quantile normalization for accurate dna methylation estimation. Epigenomics, 4(3):325–341, 2012.">TT12</a>]</span>,</p></li>
<li><p>Qunatile: Quantile normalization method <span id="id10">[<a class="reference internal" href="markdown.html#id51" title="Jovana Maksimovic, Lavinia Gordon, and Alicia Oshlack. Swan: subset-quantile within array normalization for illumina infinium humanmethylation450 beadchips. Genome biology, 13(6):1–12, 2012.">MGO12</a>]</span>,</p></li>
<li><p>Noob: Noob preprocessing <span id="id11">[<a class="reference internal" href="markdown.html#id52" title="Timothy J Triche Jr, Daniel J Weisenberger, David Van Den Berg, Peter W Laird, and Kimberly D Siegmund. Low-level processing of illumina infinium dna methylation beadarrays. Nucleic acids research, 41(7):e90–e90, 2013.">TJWVDB+13</a>]</span>,</p></li>
<li><p>Illumina: llumina preprocessing, as performed by Genome Studio (reverse engineered by minfi authors) <span id="id12">[<a class="reference internal" href="markdown.html#id53" title="Martin J Aryee, Andrew E Jaffe, Hector Corrada-Bravo, Christine Ladd-Acosta, Andrew P Feinberg, Kasper D Hansen, and Rafael A Irizarry. Minfi: a flexible and comprehensive bioconductor package for the analysis of infinium dna methylation microarrays. Bioinformatics, 30(10):1363–1369, 2014.">AJCB+14</a>]</span></p></li>
<li><p>Funnorm: Funcational normalization <span id="id13">[<a class="reference internal" href="markdown.html#id7" title="Jean-Philippe Fortin, Aurélie Labbe, Mathieu Lemire, Brent W Zanke, Thomas J Hudson, Elana J Fertig, Celia MT Greenwood, and Kasper D Hansen. Functional normalization of 450k methylation array data improves replication in large cancer studies. Genome biology, 15(11):1–17, 2014.">FLL+14</a>]</span></p></li>
</ul>
</li>
<li><p><em>Select filteration method</em>: In this section, we assigned options for the user to perform the different filters, like removal of XY chormosomes from the analysis, removal of SNPs or removal of non-specific probes from the dataset. By default, the pipeline will use p-value detection 0.01.</p></li>
<li><p><em>Compute cell type heterogeneity</em>: similar as ChAMP, we used Houseman method to correct the cell type heterogeneity. In minfi pipeline we used default minfi function <em>estimateCellCounts</em>. As before, user can choose to avoid this if the samples are not from PBMC cell types.</p></li>
<li><p><em>Please select the compare groups</em>: In this option, the pipeline will ask the user to type the group names which will be compared for the differential methylation analysis. To compute the ContrastsMatrix, these group names will be used. Please use the same groups as provide in the Sample_sheet.csv.</p></li>
<li><p><em>Choose genome annotation database</em>: To annotate the DMC list from the analysis, use the human genome reference annotation data file. For 850K array, make sure to use the hg38 array to compare the result with the output of ChAMP pipeline.
<img alt="minfi" src="_images/methylysis3.png" /></p></li>
</ol>
</div>
<div class="section" id="technical-setup">
<h6><span class="section-number">1.1.1.2.1.3. </span>Technical setup:<a class="headerlink" href="#technical-setup" title="Permalink to this headline">¶</a></h6>
<ol class="simple">
<li><p><em>Number of cores</em>: Both pipeline can run on 1 core which will take more time to compute the entire process. User can choose to setup the number of cores depending the availability. Please select maximum (n-1) cores (where n is the number of total cores in the computer).</p></li>
<li><p><em>Data upload</em>
The user should set the parameters first and choose all parameters as described above and then upload the data directory. As soon as the pipeline finishes the upload of the data directory, it will start running the analysis.
<img alt="data upload" src="_images/methylysis4.png" /></p></li>
</ol>
</div>
</div>
<div class="section" id="requirement-for-data-upload">
<h5><span class="section-number">1.1.1.2.2. </span>Requirement for data upload<a class="headerlink" href="#requirement-for-data-upload" title="Permalink to this headline">¶</a></h5>
<ol class="simple">
<li><p>idat files: all idat files, green and red as received from Illumina sequencing array should be provided for the analysis. All files should be in one directory/folder.</p></li>
<li><p>Sample_sheet.csv: the “Sample_sheet.csv” should also be provided in the same directory with idat files and</p></li>
</ol>
</div>
</div>
</div>
</div>
<div class="section" id="analysis-result">
<h2><span class="section-number">1.2. </span>Analysis Result<a class="headerlink" href="#analysis-result" title="Permalink to this headline">¶</a></h2>
<p>We implemented two different appraoches to save/view the analysis result.
A. First, all results from the analysis pipeline will be saved in the same folder from where the data loaded, and
B. Second, using the on-screen visualization tool.</p>
<p>In the visaulization tabs of methylysis tool, we introduced 4 tabs for differential methylation analysis result,</p>
<ol class="simple">
<li><p><strong>QC result</strong>: QC result from each pipeline mentioned above will be generated and presented here in PDF format. User can download the PDF, however the result will also be saved in the data directory.</p></li>
<li><p><strong>Normalized data table</strong>: Only first 100 rows from the normalized \beta values will appear in the visualization tab. The full result will be saved on the data directory.</p></li>
<li><p><strong>Cell type deconvolution plot</strong>: If user choose to analyze the cell type deconvolution, the analysis figure will appear here. Please note, current version only supports the PBMC deconvolution.</p></li>
<li><p><strong>DMC table</strong>: First 100 rows of the differentially methylated CpGs will appear here and the full result will be saved as mentioned above.</p></li>
</ol>
</div>
<div class="section" id="r-packages-used">
<h2><span class="section-number">1.3. </span>R packages used<a class="headerlink" href="#r-packages-used" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>ChAMP bioconductor</p></li>
<li><p>minfi bioconductor</p></li>
<li><p>IlluminaHumanMethylation450kanno.ilmn12.hg19</p></li>
<li><p>IlluminaHumanMethylationEPICanno.ilm10b5.hg38</p></li>
<li><p>IlluminaHumanMethylation450kmanifest</p></li>
<li><p>RColorBrewer</p></li>
<li><p>missMethyl</p></li>
<li><p>matrixStats</p></li>
<li><p>minfiData</p></li>
<li><p>Gviz</p></li>
<li><p>DMRcate</p></li>
<li><p>stringr</p></li>
<li><p>RCurl</p></li>
<li><p>Biobase</p></li>
</ol>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="intro.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Welcome to methylR: DNA Methylation Data Analysis Pipeline</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="multid.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">2. </span>Multi-D Analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Jyotirmoy Das<br/>
    
        &copy; Copyright 2022.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>